src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -284,27 +285,36 @@ def visit_Call(self, node: foast.Call, **kwargs) -> foast.Call:
                 func=new_func, args=new_args, kwargs={}, location=node.location, type=new_type
             )
         elif isinstance(new_func.type, ct.FunctionType):
-            new_args = self.visit(node.args, **kwargs)
-            new_kwargs = self.visit(node.kwargs, **kwargs)
-            try:
-                type_info.is_callable(
-                    new_func.type,
-                    with_args=[arg.type for arg in new_args],
-                    with_kwargs={keyword: arg.type for keyword, arg in new_kwargs.items()},
-                    raise_exception=True,
+            return_type = new_func.type.returns
+
+            # todo(tehrengruber): solve in a more generic way, e.g. using
+            #  parametric polymorphism.
+            # resolve polymorphic builtins
+            if not type_info.is_concrete(return_type) and node.func.id in FUN_BUILTIN_NAMES:
+                visitor = getattr(self, f"_visit_{node.func.id}")
+                return visitor(node, **kwargs)
+            else:
+                new_args = self.visit(node.args, **kwargs)
+                new_kwargs = self.visit(node.kwargs, **kwargs)
+                try:
+                    type_info.is_callable(
+                        new_func.type,
+                        with_args=[arg.type for arg in new_args],
+                        with_kwargs={keyword: arg.type for keyword, arg in new_kwargs.items()},
+                        raise_exception=True,
+                    )
+                except GTTypeError as err:
+                    raise FieldOperatorTypeDeductionError.from_foast_node(
+                        node, msg=f"Invalid argument types in call to '{node.func.id}'!"
+                    ) from err
```

Comment:
I think it makes sense to run this also for the `node` returned in the if branch.